src/intelExt.cpp

```diff
@@ -0,0 +1,184 @@
+// Copyright (C) 2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+
+/*
+ * Some simple wrappers to enable HElib to use HEXL.
+ */
+
+#ifdef USE_INTEL_HEXL
+
+#include "intelExt.h"
+
+#include <hexl/hexl.hpp>
+
+#include <unordered_map>
+#include <functional>
+#include <shared_mutex>
+#include <mutex>
+
+namespace intel {
+
+using intel::hexl::NTT;
+
+static std::shared_mutex table_mutex;
+
+struct Key
+{
+  uint64_t degree;
+  uint64_t q;
+
+  bool operator==(const Key& other) const
+  {
+    return this->degree == other.degree && this->q == other.q;
+  }
+};
+
+struct Hash
+{
+  size_t operator()(const Key& key) const
+  {
+    return std::hash<uint64_t>{}(key.degree) ^
+           (std::hash<uint64_t>{}(key.q) << 1);
+  }
+};
+
+// Lookup table to avoid re-creating previously created NTTs
+// For life of program.
+static std::unordered_map<Key, NTT, Hash> table;
+
+static NTT& initNTT(uint64_t degree, uint64_t q)
+{
+  Key key = {degree, q};
+    
+  { // First Read to see if there is an NTT
+    std::shared_lock read_lock(table_mutex);
+    auto it = table.find(key);
+    if (it != table.end()) { // Found
+      return it->second;
+    } 
+  }
+
+  { // Didn't find an NTT, lock the table for writing
+    std::scoped_lock write_lock(table_mutex);
+    // Check again to see another thread snuck it in.
+    auto it = table.find(key);
+    if (it != table.end()) { // Found
+      return it->second;
+    } else { 
+      auto ret = table.emplace(key, NTT(degree, q));
+      return (ret.first)->second; // The NTT object just created.
+    }
+  }
+}
+
+void FFTFwd(long* output,
+            const long* input,
+            long n,
+            long q) 
+{
+  initNTT(/*degree=*/n, /*modulus=*/q)
+      .ComputeForward(reinterpret_cast<uint64_t*>(output),
+                      reinterpret_cast<const uint64_t*>(input),
+                      /*input_mod_factor=*/4,
```

Comment:
Does HELib ever allow values to grow larger than the modulus `q`? If not, `input_mod_factor=1` may be faster here and with the InvNTT.